### ISSUE TITLE
As a blockchain developer I would like to use a between clause in queries

### ISSUE THREAD

**Issue by jdockter** [hyperledger/composer#2287]
In a query I would like to only return specific assets that are limited using between clause, ie assets created between two dates

## Expected Behavior
If I have 10 assets that were created over 5 days (Monday to Friday), 2 each day, if I query for assets on just Monday and Tuesday I should only get back 4 assets.

## Actual Behavior
Not possible today

## Steps to Reproduce

1.
2.
3.
4.

## Existing issues

- [ ] [Stack Overflow issues](http://stackoverflow.com/tags/hyperledger-composer)
- [ ] [GitHub Issues](https://github.com/hyperledger/composer/issues)
- [ ] [Rocket Chat history](https://chat.hyperledger.org/channel/composer)

## Your Environment

* Version used:
* Environment name and version (e.g. Chrome 39, node.js 5.4):
* Operating System and version (desktop or mobile):
* Link to your project:

**Comment by kletkeman** [comment]
You are no doubt asking for the ability to query for the first time an asset key is written into world state. I'm not sure what that would look like, but perhaps you might consider adding "DateCreated" as a property of your assets and then you can create a named query like this:

```
query readAssetHistory {
    description: "Return a list of assets by creation date"
    statement:
        SELECT org.poc.maximo.SR_HIST
            WHERE ((dateCreated >= _$date_created_start) AND (dateCreated < _$date_created_end))
}
```
Note that the start is inclusive and end is exclusive so that nanoseconds do not have to be specified. And obviously it will be necessary to index dateCreated in couchdb if you want decent performance.

**Comment by fenglianxu** [comment]
@jdockter Does @kletkeman answer your question?  You can use **and** operator to achieve this. 
Can you confirm so that we can close this story.

**Comment by inopia** [comment]
@kletkeman @fenglianxu I tried `WHERE ((dateCreated >= _$date_created_start) AND (dateCreated < _$date_created_end))` but the query ignores the `_$date_created_end` clause.

**Comment by kletkeman** [comment]
Yeah ... I think I made a mistake in my syntax. You need and underscore in front of you variables as well ... as in _dateCreated ... at least, that's how my own queries look ... sorry about that.

**Comment by inopia** [comment]
It just throws a syntax error if I put underscore before model field. I just can't get query by date range to work.

Using the 1.1.0-preview docker images for hyperledger-fabric and composer 0.17.6 btw.

```
asset Item identified by itemId {
  o String  status
  o TimeDate created
}

query getItemByDateRange{
    description: "Select all ltems of given status between given datetime range."
    statement:
        SELECT Item
            WHERE (
                (created >= _$startTime) AND
                (created < _$endTime) AND
                (status == _$status))
}
```

Just returns everything after `_$startTime` with the `_$status`.

**Comment by inopia** [comment]
I inspected the network traffic and this is the actual CouchDB query resulting from above:
```
{
    "skip": 0, 
    "limit": 10000, 
    "selector": {
        "data.$registryType": "Asset", 
        "data.created": {
            "$gte": "2018-01-01"
        }, 
        "data.$registryId": "Item", 
        "data.$class": "Item", 
        "data.status": {
            "$eq": "SHIPPED"
        }
    }
}
```

**Comment by andrew-coleman** [comment]
This should be resolved by #3694. Please reopen if you believe otherwise.  Thanks.
